src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -0,0 +1,559 @@
+            return -1;
+        }
+    }
+
+    public int getWorstCaseFrequencyDepartureTime (TripSchedule schedule, int stopPositionInPattern, int frequencyEntryIdx, int earliestTime) {
+        int headway = schedule.headwaySeconds[frequencyEntryIdx];
+        int travelTimeFromStartOfTrip = schedule.departures[stopPositionInPattern];
+        // The last vehicle could leave the terminal as early as headwaySeconds before the end of the frequency entry.
+        int earliestEndTimeOfFrequencyEntry = schedule.endTimes[frequencyEntryIdx] - headway + travelTimeFromStartOfTrip;
+
+        if (earliestEndTimeOfFrequencyEntry < earliestTime) return -1;
+
+        // board pessimistically assuming the entry is already running
+        int latestBoardTimeAssumingEntryIsAlreadyRunning = earliestTime + headway;
+        // figure out the latest departure time of this trip at this stop
+        int latestBoardTimeOfFirstTrip = schedule.startTimes[frequencyEntryIdx] + headway + travelTimeFromStartOfTrip;
+        // return the max of those two
+        return Math.max(latestBoardTimeAssumingEntryIsAlreadyRunning, latestBoardTimeOfFirstTrip);
+    }
+
+    private void doTransfers (RaptorState state) {
+        // avoid integer casts in tight loop below
+        int walkSpeedMillimetersPerSecond = (int) (request.walkSpeed * 1000);
+        int maxWalkMillimeters = (int) (request.walkSpeed * request.maxWalkTime * 60 * 1000);
+
+        for (int stop = state.nonTransferStopsTouched.nextSetBit(0); stop > -1; stop = state.nonTransferStopsTouched.nextSetBit(stop + 1)) {
+            // no need to consider loop transfers, since we don't mark patterns here any more
+            // loop transfers are already included by virtue of those stops having been reached
+            TIntList transfersFromStop = transit.transfersForStop.get(stop);
+            if (transfersFromStop != null) {
+                for (int stopIdx = 0; stopIdx < transfersFromStop.size(); stopIdx += 2) {
+                    int targetStop = transfersFromStop.get(stopIdx);
+                    int distanceToTargetStopMillimeters = transfersFromStop.get(stopIdx + 1);
+
+                    if (distanceToTargetStopMillimeters < maxWalkMillimeters) {
+                        // transfer length to stop is acceptable
+                        int walkTimeToTargetStopSeconds = distanceToTargetStopMillimeters / walkSpeedMillimetersPerSecond;
+                        int timeAtTargetStop = state.bestNonTransferTimes[stop] + walkTimeToTargetStopSeconds;
+
+                        if (walkTimeToTargetStopSeconds < 0) {
+                            LOG.error("Negative transfer time!!");
+                        }
+
+                        state.setTimeAtStop(targetStop, timeAtTargetStop, -1, stop, 0, 0, true);
+                    }
+                }
+            }
+        }
+    }
+
+    /** Get a list of the internal IDs of the patterns touched using the given index (frequency or scheduled) */
```

Comment:
Clarify what it means for a pattern to be "touched": the stop was updated in the search round that produced the state, because a faster way was found to reach it. So we're getting a bitset of all patterns that contain stops whose travel time was improved (updated) in the round that produced the supplied state.